pkg/controllers/managementuser/snapshotbackpopulate/snapshotbackpopulate.go

```diff
@@ -95,26 +118,40 @@ func (h *handler) OnChange(key string, configMap *corev1.ConfigMap) (runtime.Obj
 		return configMap, fmt.Errorf("error while listing existing etcd snapshots for cluster %s: %w", cluster.Name, err)
 	}
 
+	// currentEtcdSnapshotsToKeep is a map of etcd snapshots snapshot objects we have retrieved from the management cluster.
 	currentEtcdSnapshotsToKeep := map[string]*rkev1.ETCDSnapshot{}
-	// iterate over the current etcd snapshots
-	// if the current etcd snapshot is NOT found in the actual etcd snapshots list, we delete it
-	// otherwise, we add it to the desired etcd snapshots map
+
+	// iterate over the current etcd snapshots objects retrieved from the management cluster
+	// if snapshot object is not found in the etcd snapshot configmap, mark it missing. if it is a local snapshot
+	// and no machine can be found for it, go ahead and delete it.
+	// if the snapshot object is found in the configmap, add it to the currentEtcdSnapshotsToKeep for reconciliation
 	for _, existingSnapshotCR := range currentEtcdSnapshots {
-		storageLocation, ok := existingSnapshotCR.GetAnnotations()[StorageLabelKey]
+		storageLocation, ok := existingSnapshotCR.GetAnnotations()[StorageAnnotationKey]
 		if !ok {
 			storageLocation = StorageLocal
 			if existingSnapshotCR.SnapshotFile.NodeName == StorageS3 {
 				storageLocation = StorageS3
 			}
 		}
 		snapshotKey := existingSnapshotCR.SnapshotFile.Name + storageLocation
+		// check to see if the snapshot CR we have is found in the etcd-snapshots configmap, and if it is not found in the configmap, mark it as missing
 		if _, ok := actualEtcdSnapshots[snapshotKey]; !ok {
+			// the snapshot custom resource we are processing does not exist in the configmap
 			if storageLocation != StorageS3 {
 				// check to make sure the machine actually exists for the snapshot
-				listSuccessful, machine, err := rke2.GetMachineFromNode(h.machineCache, existingSnapshotCR.SnapshotFile.NodeName, cluster)
+				var listSuccessful bool
+				var machine *capi.Machine
+
+				if existingSnapshotCR.Labels[rke2.MachineIDLabel] == "" {
+					logrus.Debugf("[snapshotbackpopulate] rkecluster %s/%s: snapshot %s/%s was missing machine ID label: %s", cluster.Namespace, cluster.Name, existingSnapshotCR.Namespace, existingSnapshotCR.Name, rke2.MachineIDLabel)
+					// If the machineID label was not set, fall back to looking up the machine by node name, as this may be a snapshot from an earlier version of Rancher that created local snapshots using the snapshotbackpopulate controller, which means the snapshot should not have the machine ID label.
+					listSuccessful, machine, err = rke2.GetMachineFromNode(h.machineCache, existingSnapshotCR.SnapshotFile.NodeName, cluster)
```

Comment:
Is this ok to do here since the `NodeName` might change? Is there an alternative way to looking up the machine in the list?

Reply:
This is for backwards compatibility. The alternative way to looking up the machine is 2 lines below,where we look up the machine by the machine ID label on the snapshot.

Reply:
Thanks!